make http/2 settings configurable (#988)

jwfang opened the pull request:
for #988 

WIP, for your preview

jwfang commented:
i will rename CodecOptions to Http2Options,
and the old codec_options will keep its name.

seems this is only used by http2, should we move it to Http2 namespace ?

mattklein123 commented on include/envoy/http/codec.h:
```diff
@@ -150,11 +150,17 @@ class ConnectionCallbacks {
  * A list of options that can be specified when creating a codec.
  */
-class CodecOptions {
+class CodecOptionFlags {
```
This whole thing is a mess and will be cleaned up as part of the v2 API effort. cc @htuch @PiotrSikora 

In the interim, I think it's OK to leave this in the http namespace.

mattklein123 commented on include/envoy/http/codec.h:
```diff
@@ -150,11 +150,17 @@ class ConnectionCallbacks {
 };
 
+struct CodecOptions {
+  uint64_t flag_options;
```
nit: all members should end with '_'

mattklein123 commented on include/envoy/http/codec.h:
```diff
@@ -150,11 +150,17 @@ class ConnectionCallbacks {
-class CodecOptions {
+class CodecOptionFlags {
 public:
   static const uint64_t NoCompression = 0x1;
```
It's not part of your change, but as long as you are in here, can you please rename ```NoCompression``` to ```DisableDynamicHPACKTable```. Don't bother changing the actual JSON config. We will fix that in v2 API.

mattklein123 commented on include/envoy/upstream/upstream.h:
```diff
@@ -246,7 +247,7 @@ class ClusterInfo {
    *         @see Http::CodecOptions.
    */
-  virtual uint64_t httpCodecOptions() const PURE;
+  virtual Http::CodecOptions httpCodecOptions() const PURE;
```
return ```const Http::CodecOptions&``` and fix doc comment above.

mattklein123 commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -506,12 +506,16 @@ void ConnectionImpl::sendPendingFrames() {
+
+  if (codec_options.flag_options & CodecOptionFlags::NoCompression) {
     iv.push_back({NGHTTP2_SETTINGS_HEADER_TABLE_SIZE, 0});
     conn_log_debug("disabling header compression", connection_);
```
can you fix this comment to to read: "setting HPACK dynamic table size to 0"

jwfang commented:
i did the following to building envoy:

```
[vagrant@localhost ~]$ sudo docker pull lyft/envoy-build
Using default tag: latest
latest: Pulling from lyft/envoy-build
aafe6b5e13de: Already exists
0a2b43a72660: Already exists
18bdd1e546d2: Already exists
8198342c3e05: Already exists
f56970a44fd4: Already exists
10e7822aad27: Already exists
2f4f7cb8f162: Already exists
96778b1d5664: Already exists
Digest: sha256:b3196cd595fae6cc2e5f887917dbd63f95389d6a484b17f02e8b081c08c45b9a
Status: Image is up to date for lyft/envoy-build:latest

[vagrant@localhost ~]$ sudo docker run -ti --rm -v /home/vagrant/tmp/envoy-build-lyft:/build -v /home/vagrant/src/lyft/envoy:/source lyft/envoy-build bash
root@7d1b733ce347:/# cd /source/
root@7d1b733ce347:/source# ./ci/do_ci.sh bazel.dev
```

htuch commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,28 @@ class ConnectionCallbacks {
+  uint32_t max_concurrent_streams_;
+  uint32_t initial_window_size_;
+
+  static const uint32_t DEFAULT_MAX_CONCURRENT_STREAMS = 1024;
```
Nit: just a minor style note while looking at this diff, it would be good to make the capitalization more consistent for the static consts in this header (below is all caps, above is camel case).

jwfang replied:
these are copied from the old codes. :)
should i go all caps or camel case ?

mattklein123 replied:
We don't really have consistent style here unfortunately. I would probably s/DisableDynamicHPACKTable/DISABLE_DYNAMIC_HPACK_TABLE

mattklein123 commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,28 @@ class ConnectionCallbacks {
+    static const uint64_t DisableDynamicHPACKTable = 0x1;
+  };
+
+  Http2Settings()
```
nit: drop constructor definition and just do inline init below. E.g.,
```uint64_t codec_options_{0};```

etc

jwfang replied:
didn't know that could be done, from pre-2010 c++ age. thanks

mattklein123 commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,28 @@ class ConnectionCallbacks {
 
 /**
- * A list of options that can be specified when creating a codec.
+ * http/2 settings
```
nit: "http/2 codec settings"

mattklein123 commented on include/envoy/upstream/upstream.h:
```diff
@@ -243,10 +244,10 @@ class ClusterInfo {
   /**
-   * @return uint64_t HTTP codec options for HTTP connections created on behalf of this cluster.
-   *         @see Http::CodecOptions.
+   * @return Http::Http2Settings for http/2 connections created on behalf of this cluster.
```
nit: ```const Http::Http2Settings&```

jwfang replied:
sorry, it's too obvious

mattklein123 commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -522,7 +522,9 @@ void ConnectionImpl::sendSettings(uint64_t codec_options) {
-                                    DEFAULT_WINDOW_SIZE - NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE);
+                                    http2_settings.initial_window_size_ -
+                                        NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE);
+  // TODO: ensure mimium initial_window_size from json schema
```
yup

jwfang replied:
exactly what i am going to do.

mattklein123 commented on test/mocks/upstream/cluster_info.h:
```diff
@@ -28,7 +28,7 @@ class MockClusterInfo : public ClusterInfo {
   MOCK_CONST_METHOD0(perConnectionBufferLimitBytes, uint32_t());
   MOCK_CONST_METHOD0(features, uint64_t());
-  MOCK_CONST_METHOD0(httpCodecOptions, uint64_t());
+  MOCK_CONST_METHOD0(http2Settings, const Http::Http2Settings&());
```
The tests are most likely failing because you need to setup a default ON_CALL return  a default Http2Settings object. You can verify by running the tests locally with ```--test_arg="--gmock_verbose=info"```

jwfang replied:
by tests failing, do you mean unit test or integration tests?
it seems all the orignal unit tests run fine, but integration tests keep failing. i guess you already saw my earier comment on this.

not familiar with gtest/gmock, will try as you suggested and invest it.

jwfang replied:
thank you @mattklein123 , i fixed the integration tests failures by add ON_CALL.

jwfang commented:
can some explained these to me ? it seems both wired and interesting.
https://github.com/jwfang/envoy/blob/issue988/test/common/http/utility_test.cc#L19-L21

i saw Envoy::Router::RetryPolicy and its tests don't need declarations like this.

jwfang commented on test/common/http/utility_test.cc:
```diff
@@ -16,7 +16,9 @@ namespace Envoy {
 
 // Satisfy linker
-const uint64_t CodecOptions::NoCompression;
+const uint64_t Http2Settings::CodecOptions::DISABLE_DYNAMIC_HPACK_TABLE;
```
need these, else link error with undefined symbols.

but why ？

dnoe replied:
Without this there is no storage actually assigned to store this static variable.  That may be OK: if you never require actual storage for the constant int the compiler doesn't complain.  But if the static const is passed by reference or it's address taken at some point in the code then it needs an actual storage location (if it's passed by reference someone might take the address later).  This means it must be _defined_ (it's _declared_ in the header) in a single CPP file.

Why does it matter here and not in the Envoy executable? In the Envoy executable case they're never passed by reference so the compiler lets you get away with not defining the variable.

jwfang replied:
got it, thank you for detailed explanation.

dnoe replied:
Yup, that looks like it.  Good find.

dnoe commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,24 @@ class ConnectionCallbacks {
-class CodecOptions {
-public:
-  static const uint64_t NoCompression = 0x1;
+struct Http2Settings {
```
My personal preference would be to make this a class that hides a bit of it's own logic - I touched on this a bit in another comment.  It would enable writing code like calling "http2settings.use_compression()" to find out if compression is enabled rather than requiring the caller to understand that codec_options_ is really a bit field.

jwfang replied:
i remove the CodecOptions entirely, and add a new hpack_table_size config.

jwfang replied:
if this is acceptable, then i will finish the docs.

dnoe commented on test/common/http/utility_test.cc:
```diff
@@ -90,21 +90,22 @@ TEST(HttpUtility, createSslRedirectPath) {
         Json::Factory::loadFromString("{\"http_codec_options\": \"no_compression\"}");
-    EXPECT_EQ(CodecOptions::NoCompression, Utility::parseCodecOptions(*json));
+    EXPECT_EQ(Http2Settings::CodecOptions::DisableDynamicHPACKTable,
+              Utility::parseHttp2Settings(*json).codec_options_);
```
I think there's more to check here.  The other values should be set to their defaults?

dnoe commented on test/common/http/utility_test.cc:
```diff
@@ -90,21 +90,22 @@ TEST(HttpUtility, createSslRedirectPath) {
   {
     Json::ObjectSharedPtr json = Json::Factory::loadFromString("{\"http_codec_options\": \"foo\"}");
-    EXPECT_THROW(Utility::parseCodecOptions(*json), EnvoyException);
+    EXPECT_THROW(Utility::parseHttp2Settings(*json), EnvoyException);
```
You can use EXPECT_THROW_WITH_MESSAGE here, this will make sure it's not hitting some other code that throws EnvoyException.

dnoe commented on test/common/http/http2/codec_impl_test.cc:
```diff
@@ -236,8 +245,13 @@ TEST_P(Http2CodecImplTest, DeferredReset) {
+                                           Http2Settings::DEFAULT_INITIAL_WINDOW_SIZE),
+                    Http2SettingsTestParam(Http2Settings::CodecOptions::DisableDynamicHPACKTable,
+                                           Http2Settings::DEFAULT_MAX_CONCURRENT_STREAMS,
+                                           Http2Settings::DEFAULT_INITIAL_WINDOW_SIZE)));
```
You could expand the testing by also parameterizing this test case with the edge cases of valid values for max_concurrent_streams and default_initial_window_size.

jwfang commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -506,14 +506,15 @@ void ConnectionImpl::sendPendingFrames() {
 
-  if (codec_options & CodecOptions::NoCompression) {
+  if (http2_settings.codec_options_ &
+      Http::Http2Settings::CodecOptions::DISABLE_DYNAMIC_HPACK_TABLE) {
```
yes, since we are using a struct now,  we can add additional fields easily.
so the flags as bool no needed any more.

jwfang commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,25 @@ class ConnectionCallbacks {
+
+  static const uint32_t MIN_MAX_CONCURRENT_STREAMS = 1;
+  static const uint32_t DEFAULT_MAX_CONCURRENT_STREAMS = 1024;
+  static const uint32_t MAX_MAX_CONCURRENT_STREAMS = 1 << 29; // from MAX_STREAMS
```
should we change MAX_STREAMS from source/common/http/http2/conn_pool.h to using this ?

jwfang commented:
> A few comments to start. Is the config actually validated against the schema minimum and maximum?
> 
> It would be good to see some more tests that exercise the feature and corner cases (try some invalid values, etc)

i didn't do the schema check for the following reason:
1. i guess it's the json parser's job to ensure schema validation;
2. config schema test are in test/common/json/config_schemas_test_data, and they're checking against
    more complex cases, i don't want to pollute them.

but if they are needed, i will add then quickly.

dnoe commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,25 @@ class ConnectionCallbacks {
+
+  static const uint32_t MIN_INITIAL_WINDOW_SIZE = (1 << 16) - 1; // from HTTP/2 spec
+  static const uint32_t DEFAULT_INITIAL_WINDOW_SIZE = 256 * 1024 * 1024;
+  // ( 1 << 31 ) - 1 cause overflow warning, so hard-code the value
```
You can avoid the overflow by specifying the 1 as an unsigned long long literal:

`(1ULL << 31) - 1`

jwfang replied:
thanks. i will do it tomorrow.

htuch commented on source/common/http/utility.cc:
```diff
@@ -136,17 +136,27 @@ bool Utility::isInternalRequest(const HeaderMap& headers) {
+
   std::string options = config.getString("http_codec_options", "");
   for (const std::string& option : StringUtil::split(options, ',')) {
     if (option == "no_compression") {
```
This feature should be marked `DEPRECATED` in comments. Also we should track this in #1012.

jwfang replied:
see below. 
comment about this in #1012 .

htuch commented on source/common/http/utility.cc:
```diff
@@ -136,17 +136,27 @@ bool Utility::isInternalRequest(const HeaderMap& headers) {
 
+  Json::ObjectSharedPtr http2_settings = config.getObject("http2_settings", true);
+  ret.hpack_table_size_ = http2_settings->getInteger(
+      "hpack_table_size",
```
Should add a check that this isn't set non-zero at the same time as "no_compression" set, otherwise you get strange behavior.

jwfang replied:
i made 'http2_settings.hpack_table_size' silently overrides 'http_codec_options.no_compression'
here to save me some checking logic, and thought that overrides may also be ok.

i changed it to error out if the two configs conflicts. 

But i can't log warning DEPRECATED yet, since i have somehow get a logger. 
I can pass in a logger from calling site, but may lose the CONFIG logger enum info.

htuch commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,24 @@ class ConnectionCallbacks {
+
+  static const uint32_t MIN_HPACK_TABLE_SIZE = 0;
+  static const uint32_t DEFAULT_HPACK_TABLE_SIZE = 4 * 1024; // from nghttp2
+  static const uint32_t MAX_HPACK_TABLE_SIZE = 16 * 1024 * 1024;
```
Where does this come from?

jwfang replied:
comments added

jwfang commented on test/common/http/http2/codec_impl_test.cc:
```diff
@@ -25,7 +25,16 @@ using testing::NiceMock;
+struct Http2SettingsTestParam : public Http2Settings {
+  Http2SettingsTestParam(uint32_t hpack_table_size, uint32_t max_concurrent_streams,
+                         uint32_t initial_window_size) {
+    hpack_table_size_ = hpack_table_size;
```
docs added.

jwfang commented on docs/configuration/http_conn_man/http_conn_man.rst:
```diff
@@ -97,6 +98,27 @@ http_codec_options
+    range from 1 to 536870912 (2^29) and defaults to 1024.
+
+  initial_window_size
+    *(optional, integer)* `Initial flow-control window`_ size. Valid values range from 65535
```
65535 is the initial value defined in the HTTP/2 spec. 

from http://httpwg.org/specs/rfc7540.html#rfc.section.6.9.3 ,
my understanding is that client are allowed to send up to initial window bytes before got peer's window setting. if it received a windows setting smaller, it will blocks until another WINDOW_UPDATE.
so it needs to handle WINDOW_UPDATE properly to avoid this, i am not sure if this is done in envoy/nghttp2 already, so to be safe, i don't allow value below the initial size.

jwfang commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -506,24 +506,34 @@ void ConnectionImpl::sendPendingFrames() {
-                                    DEFAULT_WINDOW_SIZE - NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE);
-  ASSERT(rc == 0);
+  ASSERT(http2_settings.initial_window_size_ >= NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE);
+  if (http2_settings.initial_window_size_ > NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE) {
```
i made http2_settings.initial_window_size_ >= NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE
in config schemas, so i just use > to compare them.

i guess they are different configs,  so it make sense to not mess them togetther.
i removed the assert and use not equal now.

jwfang commented:
thanks @dnoe @htuch for your help/reviews/efforts.
hope i have fixed all your current concerns from the review. but i am sure there're something left off. :)

htuch commented:
This LGTM once comments are addressed. @jwfang Yeah, a rename would make sense as you suggest. @ccaraman can you route to the right person on the Lyft side to provide a review/approval?

jwfang commented on docs/configuration/http_conn_man/http_conn_man.rst:
```diff
@@ -97,6 +98,27 @@ http_codec_options
+
+  max_concurrent_streams
+    *(optional, integer)* `Maximum concurrent streams`_ allowed on one HTTP/2 connection. Valid values
+    range from 1 to 536870912 (2^29) and defaults to 1024.
```
docs added

jwfang commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,24 @@ class ConnectionCallbacks {
+  static const uint32_t DEFAULT_MAX_CONCURRENT_STREAMS = 1024;
+  static const uint32_t MAX_MAX_CONCURRENT_STREAMS = 1 << 29; // from MAX_STREAMS
+
+  static const uint32_t MIN_INITIAL_WINDOW_SIZE = (1 << 16) - 1; // from HTTP/2 spec
```
comments added.

jwfang commented:
i added comments/docs for the rational behind the defaults/min/max setting values.

my remaining concern about the consts currently is the MIN_INITIAL_WINDOW_SIZE.
from the HTTP/2 spec, http://httpwg.org/specs/rfc7540.html#rfc.section.6.9.3 , about reducing window size,
it seems that if the send already send the default window size data, 

> a receiver MAY continue to process streams that exceed flow-control limits

or

> The receiver MAY instead send a RST_STREAM with an error code of FLOW_CONTROL_ERROR for the affected streams

so i originally make MIN_INITIAL_WINDOW_SIZE equal the default window size to avoid this.

but after a second thought, it seems reducing the MIN_INITIAL_WINDOW_SIZE to the spec minimum, which is 1, may also be useful in some case.

we didn't change the default setting, and user who reduce it below the default window size should know what they are doing.

so i will made another commit that relax the  MIN_INITIAL_WINDOW_SIZE to 1.

htuch commented:
Does nghttp2 do the required book keeping for flow control when you go negative with the initial window size? @goaway for H2 knowledge.

jwfang commented:
when setting initial_window_size smaller than the initial data the client sent,
`TEST_P(Http2CodecImplTest, ExpectContinueHeadersOnlyResponse)`
failed with 
> [critical][assert] assert failure: static_cast<uint64_t>(rc) == length: source/common/buffer/buffer_impl.cc:77

try to run the test under gdb with
`tools/bazel-test-gdb //test/common/http:async_client_impl_test -c dbg`
but got the following output

> For help, type "help".
> Type "apropos word" to search for commands related to "word"...
> Reading symbols from /root/.cache/bazel/_bazel_root/b570b5ccd0454dc9af9f65ab1833764d/execroot/source/bazel-out/local-dbg/bin/test/common/http/http2/codec_impl_test.runfiles/envoy/test/common/http/http2/codec_impl_test...done.
> (gdb) bt
> No stack.
> (gdb)

i am not sure if this issue only occurs in test, or standalone Envoy also suffer from this yet.

htuch commented:
You need to tell gdb "run" and see the crash before "bt"
On Fri, May 26, 2017 at 8:04 AM jwfang <notifications@github.com> wrote:

> when setting initial_window_size smaller than the initial data the client
> sent,
> TEST_P(Http2CodecImplTest, ExpectContinueHeadersOnlyResponse)
> failed with
>
> [critical][assert] assert failure: static_cast<uint64_t>(rc) == length:
> source/common/buffer/buffer_impl.cc:77
>
> try to run the test under gdb with
> tools/bazel-test-gdb //test/common/http:async_client_impl_test -c dbg
> but got the following output
>
> For help, type "help".
> Type "apropos word" to search for commands related to "word"...
> Reading symbols from
> /root/.cache/bazel/_bazel_root/b570b5ccd0454dc9af9f65ab1833764d/execroot/source/bazel-out/local-dbg/bin/test/common/http/http2/codec_impl_test.runfiles/envoy/test/common/http/http2/codec_impl_test...done.
> (gdb) bt
> No stack.
> (gdb)
>
> i am not sure if this issue only occurs in test, or standalone Envoy also
> suffer from this yet.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lyft/envoy/pull/989#issuecomment-304266431>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKaLv6wxGrqqHHjnZe4Jd0VoZb5ZZuXpks5r9r--gaJpZM4NfKqZ>
> .
>

jwfang commented:
my simplified failed test is here: https://github.com/jwfang/envoy/tree/issue988-f

i fixed the gtest EXPECT, the assert still there.

> INFO: From Testing //test/common/http/http2:codec_impl_test:                                                                                                                        [45/1399]
> ==================== Test output for //test/common/http/http2:codec_impl_test:
> [==========] Running 9 tests from 2 test cases.
> [----------] Global test environment set-up.
> [----------] 1 test from Http2CodecUtility
> [ RUN      ] Http2CodecUtility.reconstituteCrumbledCookies
> [       OK ] Http2CodecUtility.reconstituteCrumbledCookies (0 ms)
> [----------] 1 test from Http2CodecUtility (0 ms total)
> 
> [----------] 8 tests from Http2CodecImplTest/Http2CodecImplTest
> [ RUN      ] Http2CodecImplTest/Http2CodecImplTest.ExpectContinueHeadersOnlyResponse/0
> [2017-05-26 13:31:55.796][6624][critical][assert] assert failure: static_cast<uint64_t>(rc) == length: source/common/buffer/buffer_impl.cc:77

jwfang commented:
one more note:

if we try changing the initial client data length and server initial window size, one of the assertion value is **initial-client-data-length % server-initial-window-size**

htuch commented:
If it's too much effort, you can just set the minimum initial window size to the H2 default, and document the future work with `TODO`, that's fine with me.

PiotrSikora commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,32 @@ class ConnectionCallbacks {
+  // initial value from HTTP/2 spec, same as NGHTTP2_DEFAULT_HEADER_TABLE_SIZE from nghttp2
+  static const uint32_t DEFAULT_HPACK_TABLE_SIZE = 4 * 1024;
+  // a 16MiB maximum, hope it's more than enough
+  static const uint32_t MAX_HPACK_TABLE_SIZE = 16 * 1024 * 1024;
```
Where is this value coming from? RFC doesn't specify upper bound, so we shouldn't impose one.

jwfang replied:
ok, we'll leave the choice to the user.

PiotrSikora commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,32 @@ class ConnectionCallbacks {
+  // a 16MiB maximum, hope it's more than enough
+  static const uint32_t MAX_HPACK_TABLE_SIZE = 16 * 1024 * 1024;
+
+  static const uint32_t MIN_MAX_CONCURRENT_STREAMS = 1;
```
Minimum value is 0.

jwfang replied:
thanks, 0 indeed make sense here.

PiotrSikora commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,32 @@ class ConnectionCallbacks {
+  static const uint32_t DEFAULT_MAX_CONCURRENT_STREAMS = 1024;
+  // All streams are 2^31. Client/Server streams are half that. Just to be on the safe
+  // side we do 2^29. Same as MAX_STREAMS in source/common/http/http2/conn_pool.h.
+  static const uint32_t MAX_MAX_CONCURRENT_STREAMS = 1 << 29;
```
RFC doesn't specify upper bound, so we shouldn't impose one.

jwfang replied:
ok, i will use maximum unsigned 32-bit here.

PiotrSikora commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,32 @@ class ConnectionCallbacks {
+  // initial value from HTTP/2 spec, same as NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE.
+  // We only support increasing window size now, so this is also the minimum.
+  static const uint32_t MIN_INITIAL_WINDOW_SIZE = (1 << 16) - 1;
+  static const uint32_t DEFAULT_INITIAL_WINDOW_SIZE = 256 * 1024 * 1024;
```
256MB as the default initial window size? It defeats the flow control.

jwfang replied:
the 256MiB is copied from old codes here:
https://github.com/lyft/envoy/blob/a8c446aafa2af08f525dbab2b052f25834b78139/source/common/http/http2/codec_impl.h#L220

what's default do you suggest ?

jwfang replied:
256MiB is from the old code here : https://github.com/lyft/envoy/blob/a8c446aafa2af08f525dbab2b052f25834b78139/source/common/http/http2/codec_impl.h#L220 

it's used for both connection-level and stream-level window.

i added a new initial_connection_window_size_ field to seperate the two window.

PiotrSikora commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -506,24 +506,33 @@ void ConnectionImpl::sendPendingFrames() {
-      {NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS, MAX_CONCURRENT_STREAMS},
-      {NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE, DEFAULT_WINDOW_SIZE}};
+      {NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS, http2_settings.max_concurrent_streams_},
+      {NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE, http2_settings.initial_window_size_}};
```
You should check if this value is different from the default (`NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE`), since there is no point in sending it if it's the same.

jwfang replied:
do you mean check against **NGHTTP2_INITIAL_WINDOW_SIZE** ? with no **CONNECTION**

i think here we are setting the stream-level window size, connection-level window is updated below.

PiotrSikora commented on docs/configuration/http_conn_man/http_conn_man.rst:
```diff
@@ -97,6 +100,40 @@ http_codec_options
+
+  .. _Maximum Table Size: http://httpwg.org/specs/rfc7541.html#maximum.table.size
+  .. _Maximum concurrent streams: http://httpwg.org/specs/rfc7540.html#rfc.section.5.1.2
+  .. _Initial flow-control window: http://httpwg.org/specs/rfc7540.html#InitialWindowSize
```
Nit: could you change link anchors to `#rfc.section.X.X.X` to make things consistent?

jwfang replied:
done

jwfang commented:
thanks @PiotrSikora for your review.

i adjusted the min/max, and added a new *initial_connection_window_size_* field for connection-level window.

i left the **MIN_MAX_CONCURRENT_STREAMS** to **1** since it make tests fail, currently i have trouble
run it gdb. i will fixed that to 0 and the tests later.

jwfang commented on test/common/http/utility_test.cc:
```diff
@@ -90,21 +87,58 @@ TEST(HttpUtility, createSslRedirectPath) {
+    EXPECT_EQ(0, http2_settings.hpack_table_size_);
+    EXPECT_EQ(Http2Settings::DEFAULT_MAX_CONCURRENT_STREAMS,
+              http2_settings.max_concurrent_streams_);
+    EXPECT_EQ(Http2Settings::DEFAULT_INITIAL_WINDOW_SIZE, http2_settings.initial_window_size_);
```
add connection window

jwfang commented on test/common/http/http2/codec_impl_test.cc:
```diff
@@ -236,8 +275,35 @@ TEST_P(Http2CodecImplTest, DeferredReset) {
-                        testing::Values(0, CodecOptions::NoCompression));
+INSTANTIATE_TEST_CASE_P(
+    Http2CodecImplTest, Http2CodecImplTest,
+    ::testing::Combine(
```
skip default to speed up test

jwfang replied:
add a alldefault test

jwfang commented on test/common/http/http2/codec_impl_test.cc:
```diff
@@ -25,7 +26,25 @@ using testing::NiceMock;
+typedef ::testing::tuple<uint32_t, uint32_t, uint32_t, uint32_t> http2SettingsTuple;
+typedef ::testing::tuple<http2SettingsTuple, http2SettingsTuple> http2SettingsTupleTuple;
+
+Http2Settings http2SettingsFromTuple(const http2SettingsTuple& tp) {
```
static

jwfang commented on docs/configuration/http_conn_man/http_conn_man.rst:
```diff
@@ -97,6 +100,44 @@ http_codec_options
+.. _config_http_conn_man_http2_settings:
+
+http2_settings
+  *(optional, object)* Additional HTTP/2 codec options that are passed directly to the HTTP/2 codec. 
```
settings

jwfang commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,45 @@ class ConnectionCallbacks {
+
+  // prevent creation of new streams from peer
+  static const uint32_t MIN_MAX_CONCURRENT_STREAMS = 0;
+  // defaults to no limit, same as nghttp2
```
maximum

htuch commented on docs/configuration/http_conn_man/http_conn_man.rst:
```diff
@@ -97,6 +100,44 @@ http_codec_options
+
+.. _config_http_conn_man_http2_settings_initial_window_size:
+
+  initial_window_size
```
Nit: can you rename this `initial_stream_window_size` now that we also have `initial_connection_window_size`?

htuch commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,45 @@ class ConnectionCallbacks {
+  // TODO(jwfang): support other HTTP/2 settings
+  uint32_t hpack_table_size_{DEFAULT_HPACK_TABLE_SIZE};
+  uint32_t max_concurrent_streams_{DEFAULT_MAX_CONCURRENT_STREAMS};
+  uint32_t initial_window_size_{DEFAULT_INITIAL_WINDOW_SIZE};
```
Nit: `DEFAULT_INITIAL_STREAM_WINDOW_SIZE`.

htuch commented on test/common/http/http2/codec_impl_test.cc:
```diff
@@ -25,7 +26,25 @@ using testing::NiceMock;
 
-class Http2CodecImplTest : public testing::TestWithParam<uint64_t> {
+typedef ::testing::tuple<uint32_t, uint32_t, uint32_t, uint32_t> http2SettingsTuple;
+typedef ::testing::tuple<http2SettingsTuple, http2SettingsTuple> http2SettingsTupleTuple;
```
Nit: more descriptive type name than `...TupleTuple` please. Also, Please start type names with capital letters, e.g. `Http2SettingsTuple`.

jwfang replied:
done.

htuch commented on test/common/http/http2/codec_impl_test.cc:
```diff
@@ -25,7 +26,25 @@ using testing::NiceMock;
+typedef ::testing::tuple<uint32_t, uint32_t, uint32_t, uint32_t> http2SettingsTuple;
+typedef ::testing::tuple<http2SettingsTuple, http2SettingsTuple> http2SettingsTupleTuple;
+
+static Http2Settings http2SettingsFromTuple(const http2SettingsTuple& tp) {
```
Nit: slight preference to use an anonymous namespace rather than static in C++.

jwfang replied:
another nice trick. thanks.

htuch commented on test/common/http/http2/codec_impl_test.cc:
```diff
@@ -25,7 +26,25 @@ using testing::NiceMock;
+}
+
+// skip test if server-side max concurrent streams is 0
+#define SKIP_TEST_IF_NOT_SUITABLE                                                                  \
```
If you're going to make this zero (and I think it's fine to leave it as 1), then you probably do want to use gmock to verify that we get the right behavior here.

PiotrSikora commented:
@jwfang @htuch Agreed, if `MAX_CONCURRENT_STREAMS` 0 makes it harder to test and requires workarounds, then I'm fine with leaving minimum value as 1 for now.

PiotrSikora commented on include/envoy/http/codec.h:
```diff
@@ -148,11 +148,45 @@ class ConnectionCallbacks {
+  // TODO(jwfang): make this 0 to support decrease window size
+  static const uint32_t MIN_INITIAL_WINDOW_SIZE = (1 << 16) - 1;
+  // initial stream-level value from HTTP/2 spec, same as NGHTTP2_INITIAL_WINDOW_SIZE from nghttp2
+  static const uint32_t DEFAULT_INITIAL_WINDOW_SIZE = (1 << 16) - 1;
```
I've noticed that you've changed the default from current `256MB` to `64KB` resulting in 4096x smaller window size.

As previously noted, `256MB` effectively disables flow control, so I definitely agree that we should lower it, but such drastic change will result in different traffic characteristics, and therefor should be committed separately, otherwise we're risking reverting this commit as a whole.

Maybe let's leave it as `256MB` in this PR and let's change the default to `1MB` in subsequent PR?

@htuch any thoughts?

jwfang commented:
thanks @htuch @PiotrSikora .

i summarized desired min/max/default **values** changes below for the **this** PR:
1. hpack_table     
     none
2. max_concurrent_streams
     change minimum to 1
3. initial_stream_window
     change default to 256MB
4. initial_connection_window
     also default to 256MB

new PR:
change both default window size to 64KB

potential TODOs, will results in TODO comments in codes:
1. hpack_table
    none
2. max_concurrent_streams
     minimum to 0
3. initial_stream_window
     minimum to 0
4. initial_connection_window
     minimum to 0
3,4 may not make any sense unless we can increase the window back later some how, but that's another story.

htuch commented:
@jwfang SGTM. @PiotrSikora I think we'll be effectively lowering the buffer size and flow control window when implementing the buffer management proposal in #150 (which continues to exist as a P2 for me), so I'm not sure if we need to go ahead and lower it before then. I agree that if we do lower it that 64KB might be too low and have potential downside performance effects in a high bandwidth datacenter or cloud environment.

htuch commented on docs/configuration/http_conn_man/http_conn_man.rst:
```diff
@@ -86,6 +87,8 @@ add_user_agent
 .. _config_http_conn_man_http_codec_options:
 
 http_codec_options
+  **DEPRECATED**, use :ref:`http2_settings <config_http_conn_man_http2_settings>` instead.
```
Can you also update https://github.com/lyft/envoy/blob/master/DEPRECATED.md? Thanks.

jwfang replied:
done

jwfang commented on source/common/http/utility.cc:
```diff
@@ -136,12 +136,31 @@ bool Utility::isInternalRequest(const HeaderMap& headers) {
   for (const std::string& option : StringUtil::split(options, ',')) {
     if (option == "no_compression") {
-      ret |= CodecOptions::NoCompression;
+      if (http2_settings->hasObject("hpack_table_size") && ret.hpack_table_size_ != 0) {
```
it seems you're seeing an outdated version, :)

it's tested by this, if i am not misunderstanding you.

```
 +    auto json = Json::Factory::loadFromString(R"raw({
 +                                           "http_codec_options": "no_compression",
 +                                          "http2_settings" : {
 +                                            "hpack_table_size": 1
 +                                          }
 +                                        })raw");
 +    EXPECT_THROW_WITH_MESSAGE(
 +        Utility::parseHttp2Settings(*json), EnvoyException,
 +        "'http_codec_options.no_compression' conflicts with 'http2_settings.hpack_table_size'");
    }		    }
```

jwfang commented on docs/configuration/http_conn_man/http_conn_man.rst:
```diff
@@ -97,6 +100,40 @@ http_codec_options
+    the dynamic HPACK table. Valid values range from 0 to 16777216 (2^24) and defaults to 4096,
+    with 0 effectively disables header compression.
+
+    NOTE: A 16MiB (2^24) maximum table size seems to be more than enough.
```
this seems wired, the following is not there:
+    NOTE: A 16MiB (2^24) maximum table size seems to be more than enough.

i guess same problem with the test for below:
+      if (http2_settings->hasObject("hpack_table_size") && ret.hpack_table_size_ != 0) {

jwfang commented on docs/configuration/http_conn_man/http_conn_man.rst:
```diff
@@ -97,6 +100,40 @@ http_codec_options
+    *(optional, integer)* `Maximum concurrent streams`_ allowed on one HTTP/2 connection. Valid values
+    range from 1 to 536870912 (2^29) and defaults to 1024.
+
+    NOTE: Total HTTP/2 streams is 2^31, one side (client/server) is 2^30. To be safe, we use 2^29.
```
all NOTE except we doesn't support decrease window are removed

jwfang commented:
@htuch @PiotrSikora i change minimum **max_concurrent_stream** to 1 and default **stream/connection_window_size** to 256MB.

also some other changes from the review. hope nothing is left.

htuch commented:
@jwfang Thanks. I think you might have done a commit amend and force push, as I didn't see the diff appear in GH. In future, for reviewer convenience, it'd be helpful to have all fixups as distinct commits to speedup review.

htuch commented:
@ccaraman @RomanDzhabarov @junr03 Do one of you want to take a look at this?

junr03 commented on docs/configuration/http_conn_man/http_conn_man.rst:
```diff
@@ -97,6 +100,44 @@ http_codec_options
+
+  hpack_table_size
+    *(optional, integer)* `Maximum Table Size`_ (in octets) that the encoder is permitted to use for
+    the dynamic HPACK table. Valid values range from 0 to 4294967295 (2^32 - 1) and defaults to 4096,
```
defaults to 4096. 0 effectively disables header compression.

junr03 commented on docs/configuration/http_conn_man/http_conn_man.rst:
```diff
@@ -97,6 +100,44 @@ http_codec_options
+    (2^16 - 1, HTTP/2 default) to 2147483647 (2^31 - 1, HTTP/2 maximum) and defaults to 268435456
+    (256 * 1024 * 1024).
+
+    NOTE: 65535 is the initial window size from HTTP/2 spec. We only support increase the default window
```
s/increase/increasing

junr03 commented on docs/configuration/http_conn_man/http_conn_man.rst:
```diff
@@ -97,6 +100,44 @@ http_codec_options
+  Currently supported settings are:
+
+  hpack_table_size
+    *(optional, integer)* `Maximum Table Size`_ (in octets) that the encoder is permitted to use for
```
for consistency, can you change the links here to `link-name <link>`_. And delete the links at the end of this section.

jwfang commented:
@junr03 thanks for your review, docs updated.
@htuch _the commits showing under previous comment_ things indeed caused by timestamp, my VM's
date lags by days during the weekend.

htuch commented:
@jwfang Merged, thanks. Can we also close out #987?

jwfang commented:
thanks @htuch .

#987 is different from this one, the title is misleading.

this PR is to advertise our http/2 setting to our peer, #987 is for envoy to respect peer's setting.
but according to @mattklein123 :

>  If we overflow the other side, the code will correctly handle the REFUSED_STREAM reset that happens.

i am closing it  and will create a new one if really needed.

mattklein123 commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -506,24 +506,59 @@ void ConnectionImpl::sendPendingFrames() {
+         http2_settings.initial_connection_window_size_ <=
+             Http2Settings::MAX_INITIAL_CONNECTION_WINDOW_SIZE);
+
+  std::vector<nghttp2_settings_entry> iv;
```
For posterity, for perf we should switch this to a fixed size array with a pre-defined max size. Added to https://github.com/lyft/envoy/issues/210.
